helper/schema/resource.go

```diff
@@ -195,6 +195,61 @@ func (r *Resource) ShimInstanceStateFromValue(state cty.Value) (*terraform.Insta
 	return s, nil
 }
 
+// this type could also carry user config and prior state
+// should we want that exposed to the CRUD methods
+// there are potential issues that could arise from decisions
+// based on that data during the apply step (CRUD funcs)
+// (Core team explains this best)
+type applyResourceChangeRequest struct {
+	plannedState cty.Value
+	resp         chan<- cty.Value
+}
+
+// due to the potentially concurrent nature of the grpc server
+// create unique channels for each instance ID
+// to be created instances (id == "") should push to a large buffer chan
+// as seen below
+var applyResoureChangeRequests map[string]chan applyResourceChangeRequest = make(map[string]chan applyResourceChangeRequest)
+
+// ApplyResourceChange passes the planned state to the correct channel
+// it returns a channel for the new state to be passed back to the caller
+func ApplyResourceChange(resourceType string, id string, state cty.Value) <-chan cty.Value {
+	i := resourceType
+	if id != "" {
+		i = i + "." + id
+	}
+	if applyResoureChangeRequests[i] == nil {
+		size := 1
+		// the empty id channel should have a large buffer
+		// so we can process many parallel create funcs
```

Comment:
My thought was we could be receiving many `ApplyResourceChange` requests in parallel, and in the case where the instance id is empty (so this is effectively a CreateFunc call), I wouldn't want to block as we should be able to create unrelated resources in parallel.

It's tricky because `ApplyResoureChange` doesn't think of things in terms of CRUD really, and I created this map so we could pass instance scoped data (UPDATE/DELETE) to the CRUD methods, in the case of a CREATE though, the request isn't scoped to any instance yet (it doesn't exist) it is only scoped to the resource type. (sorry if that explanation made no sense)

Reply:
I have seen the light, I was able to reimplement the response with just a pointer, I also envision us having structs that mimic the grpc requests/responses (you can sorta see that now in my updated PR).

I have kept the global request map as the mechanism of passing the data to the "new CRUD funcs", only because they make convenient queues.

Reply:
I can see how the extra large buffer isn't practically helpful